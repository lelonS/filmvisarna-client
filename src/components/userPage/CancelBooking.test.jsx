import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CancelBooking from './CancelBooking.jsx';
import newDateFormat from '../../service/newDateFormat.js';

// Test data imports
import { booking } from '../../../test-mocks/index.js';

// Mocks
const mockUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUsedNavigate,
}));

// Mock the performRequest function
import { performRequest } from '../../service/fetchService';
jest.mock('../../service/fetchService');
performRequest.mockResolvedValue({ message: 'Din bokning är nu avbokad!' });

window.alert = jest.fn();

// mock the reload function
Object.defineProperty(window, 'location', {
    configurable: true,
    value: { reload: jest.fn() },
});

describe('ConfirmBooking', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('renders correctly', () => {
        render(<CancelBooking booking={booking} setToggle={jest.fn()} />);

        // Assert that the form elements are rendered correctly
        expect(screen.getByText('Är du säker på att du vill avboka?')).toBeInTheDocument();
        expect(screen.getAllByText(booking.movie.title)[0]).toBeInTheDocument();
        expect(screen.getByText(newDateFormat(booking.screening.date))).toBeInTheDocument();
        expect(screen.getByText(booking.screening.time)).toBeInTheDocument();
        expect(screen.getByText('BokningsNr: ' + booking.bookingId)).toBeInTheDocument();
        expect(screen.getByText(booking.screening.theaterName)).toBeInTheDocument();
        expect(screen.getByText(booking.rows[0].row)).toBeInTheDocument();
        expect(screen.getByText(booking.seats.map((seat) => seat.seatNumber).join(", "))).toBeInTheDocument();
        expect(screen.getByText(booking.price + 'kr')).toBeInTheDocument();
        expect(screen.getByText(booking.status === true ? 'Bokad' : 'Avbokad')).toBeInTheDocument();
        expect(screen.getByRole('button', { name: 'Avboka' })).toBeInTheDocument();
        expect(screen.getByRole('button', { name: 'Tillbaka' })).toBeInTheDocument();
    });

    test('renders correctly with snapshot', () => {
        const tree = renderer.create(<CancelBooking booking={booking} setToggle={jest.fn()} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('cancelBooking success', async () => {

        render(<CancelBooking booking={booking} setToggle={jest.fn()} />);

        const cancelButton = screen.getByRole('button', { name: 'Avboka' });

        // Simulate changing checkbox using fireEvent
        fireEvent.click(cancelButton);
        // await new Promise((r) => setTimeout(r, 1000));

        await waitFor(() => {
            expect(performRequest).toHaveBeenCalledTimes(1);
            expect(performRequest).toHaveBeenCalledWith('/api/bookings', 'PATCH', { id: booking._id });
            expect(window.location.reload).toHaveBeenCalledTimes(1);
        });
    });

    test('cancelBooking failed to unbook', async () => {

        performRequest.mockResolvedValueOnce({ message: 'Error' });

        render(<CancelBooking booking={booking} setToggle={jest.fn()} />);

        const cancelButton = screen.getByRole('button', { name: 'Avboka' });

        // Simulate changing checkbox using fireEvent
        fireEvent.click(cancelButton);
        // await new Promise((r) => setTimeout(r, 1000));

        await waitFor(() => {
            expect(performRequest).toHaveBeenCalledTimes(1);
            expect(performRequest).toHaveBeenCalledWith('/api/bookings', 'PATCH', { id: booking._id });
            expect(window.alert).toHaveBeenCalledTimes(1);
            expect(window.alert).toHaveBeenCalledWith('Något gick fel');
        });
    });


});