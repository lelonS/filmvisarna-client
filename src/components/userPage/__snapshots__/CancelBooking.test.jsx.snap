// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmBooking renders correctly with snapshot 1`] = `
<div
  className="flex flex-col items-center mb-4 mt-12 xs:max-w-[32rem] xs:mx-auto"
>
  <h4
    className="text-gold text-xl sm:text-2xl md:text-3xl mb-4 mt-6 mx-4"
  >
    Är du säker på att du vill avboka?
  </h4>
  <div
    className="flex flex-col w-5/6 xs:w-full justify-center xs:flex-row xs:items-end text-white xs:mb-6"
  >
    <img
      alt="poster of the movie: The Creator"
      className="w-[10rem] rounded-lg mx-auto xs:mx-0 mb-4 xs:mb-0"
      src="/img/theCreatorPoster.jpg"
    />
    <div
      className="hidden xs:flex flex-col gap-1 mt-4 xs:mb-0 "
    >
      <p
        className="text-md md:text-xl xs:ml-4"
      >
        The Creator
      </p>
      <p
        className="text-base md:text-lg xs:ml-4"
      >
        Söndag 24:e december
      </p>
      <p
        className="text-base md:text-lg xs:ml-4"
      >
        Klockan 
        21:00
      </p>
      <p
        className="text-base md:text-lg xs:ml-4"
      >
        BokningsNr: 
        9FA0HM
      </p>
    </div>
  </div>
  <div
    className="text-white w-4/6 xs:w-[21rem]"
  >
    <ul
      className="flex flex-col md:text-lg"
    >
      <div
        className="xs:hidden"
      >
        <li
          className="flex justify-between"
        >
           
          <p
            className="text-gold"
          >
            Film:
          </p>
           
          <p>
            The Creator
          </p>
        </li>
        <li
          className="flex justify-between"
        >
           
          <p
            className="text-gold"
          >
            BokningsNr:
          </p>
           
          <p>
            9FA0HM
          </p>
        </li>
        <li
          className="flex justify-between"
        >
           
          <p
            className="text-gold"
          >
            Datum:
          </p>
           
          <p>
            2023-12-24
          </p>
        </li>
        <li
          className="flex justify-between"
        >
           
          <p
            className="text-gold"
          >
            Tid:
          </p>
           
          <p>
            21:00
          </p>
        </li>
      </div>
      <li
        className="flex justify-between"
      >
         
        <p
          className="text-gold"
        >
          Salong:
        </p>
         
        <p>
          Stora Salongen
        </p>
      </li>
      <li
        className="flex justify-between"
      >
         
        <p
          className="text-gold"
        >
          Rad:
        </p>
         
        <p>
          8
        </p>
      </li>
      <li
        className="flex justify-between"
      >
         
        <p
          className="text-gold"
        >
          Plats:
        </p>
         
        <p>
          73, 
          74, 
          75, 
          76, 
          77, 
          78
        </p>
      </li>
      <li
        className="flex justify-between"
      >
         
        <p
          className="text-gold"
        >
          Pris:
        </p>
         
        <p>
          600
          kr
        </p>
      </li>
      <li
        className="flex justify-between"
      >
         
        <p
          className="text-gold mb-6"
        >
          Status:
        </p>
         
        <p>
          Bokad
        </p>
      </li>
      <li
        className="flex justify-evenly"
      >
         
        <button
          className=" text-lg rounded-md px-4 py-1 bg-red-600 text-white"
          onClick={[Function]}
        >
          Avboka
        </button>
        <button
          className="rounded-md text-lg px-4 py-1 bg-gold text-black"
          onClick={[Function]}
        >
          Tillbaka
        </button>
         
      </li>
    </ul>
  </div>
</div>
`;
